### PR TITLE
C51-48: Update HTML of case card component

### PR DESCRIPTION
**Pull request by swastikpareek**
## Overview
Pr covers HTML and CSS changes for the case card component on case list page and other BEM and CSS improvements

## Description.
 Following changes are done in the PR.
 * Updated HTML for case card component with BEM class structure
 * Added new CSS as per the new mockup for case card 
 * Update Loading screen to match the comps and updated class to reflect BEM structure
 * Fix some contact card style to make them more generic and reusable
 * Added Animation styles for checkbox show/hide functionality.

## Before 

c51-48 - before

## After

 * Comps 
c51-48 - updated design

 * Checkbox animation
  c51-48 checkbox animation

 * Loading styles updates
  c51-48 loading

 * Empty states
c51-48- empty states

#### Checkbox animation classes

**civicase__animated-checkbox-card** : Class to be applued to any card which will have show/hide animated checkbox
**civicase__animated-checkbox-card--expanded** : Expanded state of the card (when checkbox is visible)

#### Case card component classes

**civicase__case-card** : Class for a case-card component
**civicase__case-card--active** : Class added to case-card component which is selected for view
**civicase__case-card--closed** : Class for case-card component whose case status is closed (to have stripe bg)

**civicase__case-card__contact** : Class for case-card contact component (Used for specific styles)
**civicase__case-card__case-id** : Class for case card case id component
**civicase__case-card__type** : Class for case card case type component
**civicase__case-card__next-milestone** : Class for case card next milestone component
  **civicase__case-card__next-milestone-date** : Class for case card next milestone date component
**c civicase__case-card__activity-info ** : Class for case card info elements viz. tasks comms and other.
  **civicase__case-card__activity-count-container** : Class for case card activity counts container
    **civicase__case-card__activity-count** : Class for case card activity count component
    **civicase__case-card__activity-count--zero** : Different styles if the count is zero.
**civicase__case-card__tags-container** : Class for case card tags container
  **civicase__case-card__tag** : Class for case card specific tag 

#### Helper classes

**civicase__overdue** : Class added to add styles any element which needs red text and an exclamation button (for showing its overdue)
**civicase__pipe** : Class to style a specific pipe element

#### Update previous classes

**civicase__case-placeholder-row** changes from **case-placeholder-row** : Responsible for showing loading placeholder rows
**civicase__panel-empty** changes from **civicase-panel-empty** : Responsible for empty state of a table pane

### Other details
 *  Moved HTML for case card component HTML from `caseListTable.html` to a new file `caseCard.html` to increase modualrity
 * Old HTML code is not removed and is just commented so as to work done in C51-49 could be refered from it.
 * Increased width of case card component from `310px` to `380px` to match comps
 * CSS for `dropdown-menu` in `_contact-card.scss` is made generic
 * Created a new Helper css file to contains generic helper classes and style `components/_helper.scss`
 * Added a component `_badge` in overrides folder which overrides some of the styles of shoreditch badge element as per civicase mock ups.

### Updates

* Override default `.badges` styles to reflect comps .

### CODE REVIEW

**Comment by deb1990** [comment]
2018-09-03 at 1 10 pm
Should be of primary blue color.

File: ang/civicase/CaseListTable.html [hunk 1]
```diff
@@ -44,61 +44,14 @@
   </thead>
   <tbody>
   <tr ng-repeat="item in cases">
-    <td ng-click="viewCase(item.id, $event)" ng-class="{'case-selected-for-view': viewingCase === item.id}" class="fixed-col">
```
[the remark this answers is not in the document]
Comment (swastikpareek) [reply ang/civicase/CaseListTable.html]: The old markup is not deleted. It has been moved in `CaseCard.html` file.

File: ang/civicase/CaseListTable.html [hunk 2]
```diff
@@ -44,61 +44,14 @@
...
-             ng-class="{'item-overdue': item.activity_summary[category.name][0].is_overdue}"
-             ng-if="item.category_count.incomplete[category.name]"
-             title="{{ ts('%1 upcoming %2', {1: item.category_count.incomplete[category.name], 2: category.label}) }}"
-          >
-            <i class="fa {{ category.icon }}"></i>
-            <span class="badge noti-badge">{{ item.category_count.incomplete[category.name] }}</span>
-          </a>
-        </div>
-      </div>
+    <td ng-click="viewCase(item.id, $event)" ng-class="{'civicase__case-card--active': viewingCase === item.id}" class="civicase__case-card panel-default panel-strip-primary civicase__animated-checkbox-card" ng-include="'~/civicase/caseCard.html'">
```
[the remark this answers is not in the document]
Comment (swastikpareek) [reply ang/civicase/CaseListTable.html]: > civicase__case-card--active : Class added to case-card component which is selected for view
The class is used to specify if the case is selected for view.

File: ang/civicase/caseCard.html [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+       title="{{ ts('%1 upcoming %2', {1: item.category_count.incomplete[category.name], 2: category.label}) }}"
+    >
+      <i class="fa {{ category.icon }}"></i>
+      <span class="badge noti-badge">{{ item.category_count.incomplete[category.name] }}</span>
+    </a>
+  </div>
+</div>
+ -->
+
+ <div>
```
[the remark this answers is not in the document]
Comment (swastikpareek) [reply ang/civicase/caseCard.html]: For the CSS to work properly the `civicase__case-card` class needs to be added on the `td` element on which the directive(to be build) would apply. For that reason it is made a part of main `td` element and not this parent div.

File: css/caseList.css [hunk 4]
```diff
@@ -36,7 +36,7 @@
 #bootstrap-theme div.case-list-panel:not(.viewing-case) .scrolling table tr td:first-child {
   left: 0;
   position: absolute;  
-  width: 310px;
+  width: 380px;
```
[the remark this answers is not in the document]
Comment (swastikpareek) [reply css/caseList.css]: I was waiting for the PR for C51-38 to be merged to ignore conflicts. Will make it as a part of scss.

File: scss/components/_case-card.scss [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+  }
+
+  .civicase__case-card__next-milestone {
+    display: none;
+  }
+
+  .civicase__case-card__next-milestone-date,
+  .civicase__case-card__task-count,
+  .civicase__case-card__other-count {
+    color: inherit;
```
[the remark this answers is not in the document]
Comment (swastikpareek) [reply scss/components/_case-card.scss]: If the case is closed (have class `civicase__case-card--closed`) then to make other elements look grey we don't need to explicitly override/add color grey to them instead, we are just telling the elements to inherit the parent color which is grey which is a better approach semantically.

File: scss/components/_case-card.scss [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+    font-weight: 400;
```
[the remark this answers is not in the document]
Comment (swastikpareek) [reply scss/components/_case-card.scss]: If the case card has class `civicase__case-card--closed` then we need to make font weight normal again .

File: scss/components/_checkbox.scss [hunk 7]
```diff
@@ -14,3 +21,23 @@
   margin-left: -4px;
   margin-top: -4px;
 }
+
+.civicase__animated-checkbox-card {
```
[the remark this answers is not in the document]
Comment (swastikpareek) [reply scss/components/_checkbox.scss]: As this is closely integrated with the civicase_checbox styles. I think its better to keep them in a same file
>
```sass
.civicase__checkbox--checked {
  color: $crm-grayblue-darker;
  font-size: 24px;
  margin-left: -4px;
  margin-top: -4px;
}

.civicase__animated-checkbox-card {
  position: relative;
  transition: 0.3s padding-left cubic-bezier(0, 0, 0, 0.4);
  transition-delay: 0.2s;
}

.civicase__animated-checkbox-card--expanded {
  padding-left: 30px;
  transition-delay: 0;

  .civicase__checkbox { // here it override checkbox styles
    opacity: 1;
    transform: scale(1);
    transition-delay: 0.5s;
  }  
}
```

File: ang/civicase/CaseListTable.html [hunk 8]
```diff
@@ -124,18 +77,19 @@ <h4>{{ item.client[0].display_name }}</h4>
...
       <div style="width: 50%; margin-top: 15px;" class="oneline-loading-placeholder"></div>
     </td>
     <td ng-show="!viewingCase"><div class="activity-card-loading-placeholder"><div></div></div></td>
     <td ng-show="!viewingCase" ng-repeat="i in [1,2,3,4,5,6,7]"><div class="oneline-loading-placeholder"></div></td>
   </tr>
 
-  <tr ng-if="!cases.length && !isLoading" class="civicase-panel-empty">
-    <td class="fixed-col"></td>
+  <tr ng-if="!cases.length && !isLoading" class="civicase__panel-empty">
+    <td class="civicase__case-card panel-default panel-strip-primary"></td>
```
[the remark this answers is not in the document]
Comment (swastikpareek) [reply ang/civicase/CaseListTable.html]: It looks incorrect with current HTML. I have fixed the HTML and added the empty state design in the description.

File: scss/overrides/_badges.scss [hunk 9]
```diff
@@ -0,0 +1,6 @@
+.civicase__badge {
```
[the remark this answers is not in the document]
Comment (swastikpareek) [reply scss/overrides/_badges.scss]: We are overriding `.badge` as the styles are common site wide

File: ang/civicase/caseCard.html (hunk 3 above)
[the remark this answers is not in the document]
Comment (deb1990) [reply ang/civicase/caseCard.html]: Looks like you have moved it. Thanks

[the remark this answers is not in the document]
Comment (swastikpareek) [reply ang/civicase/caseCard.html]: Yes. after thinking out loud, it seems to be semantically better to use the approach.

File: ang/civicase/CaseCard.html [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+       title="{{ ts('%1 upcoming %2', {1: item.category_count.incomplete[category.name], 2: category.label}) }}"
+    >
+      <i class="fa {{ category.icon }}"></i>
+      <span class="badge noti-badge">{{ item.category_count.incomplete[category.name] }}</span>
+    </a>
+  </div>
+</div>
+ -->
+
+ <div class="panel-default panel-strip-primary civicase__case-card civicase__animated-checkbox-card">
```
Comment (deb1990) [inline ang/civicase/CaseCard.html]: `panel` class is still missing.

File: ang/civicase/CaseListTable.html [hunk 11]
```diff
@@ -44,61 +44,14 @@
...
-             ng-class="{'item-overdue': item.activity_summary[category.name][0].is_overdue}"
-             ng-if="item.category_count.incomplete[category.name]"
-             title="{{ ts('%1 upcoming %2', {1: item.category_count.incomplete[category.name], 2: category.label}) }}"
-          >
-            <i class="fa {{ category.icon }}"></i>
-            <span class="badge noti-badge">{{ item.category_count.incomplete[category.name] }}</span>
-          </a>
-        </div>
-      </div>
+    <td ng-click="viewCase(item.id, $event)" ng-class="{'civicase__case-card--active': viewingCase === item.id}" class="" ng-include="'~/civicase/CaseCard.html'">
```
Comment (deb1990) [inline ang/civicase/CaseListTable.html]: `ng-class="{'civicase__case-card--active': viewingCase === item.id}"` should be moved to CaseCard.html. As applying the active class is the job the directive controller, not the parent controller.

File: css/caseList.css [hunk 12]
```diff
@@ -10,7 +10,7 @@
   width: 100%;
```
Comment (deb1990) [inline css/caseList.css]: lets comment this file to make sure that we are not using any of this css.

File: css/general.css [hunk 13]
```diff
@@ -31,17 +31,17 @@
   width: 10px;
   display: inline-block;
 }
-#bootstrap-theme .civicase-panel-empty {
+#bootstrap-theme .civicase__panel-empty {
```
Comment (deb1990) [inline css/general.css]: What is the use of this `civicase__panel-empty` class? And if you are making changes, this should be moved to scss. We cannot make any changes to the existing css files.

File: scss/components/_case-card.scss [hunk 14]
```diff
@@ -0,0 +1,97 @@
...
+  min-height: 150px;
+
+  .civicase__contact-name,
+  .civicase__contact-additional__container,
+  .civicase__case-card__type {
+    text-decoration: line-through;
+  }
+
+  .civicase__case-card__next-milestone {
+    display: none;
```
Comment (deb1990) [inline scss/components/_case-card.scss]: This should not controlled from the css. For closed cases, there are no next activities thats why we dont see them. So we should do it part of the JS functionality.

File: ang/civicase/CaseCard.html [hunk 15]
```diff
@@ -0,0 +1,109 @@
...
+          </li>
+        </ul>
+      </div>
+    </div>
+    <div class="civicase__case-card__case-id">Case ID: 23</div>
+    <div class="civicase__case-card__type">Case type lorem ipsum</div>
+    <div class="civicase__case-card__next-milestone">
+      <span>Next Milestone: </span>
+      <span class="civicase__case-card__next-milestone-date civicase__overdue-activity-icon">29/01/2019</span></div>
+    <div class="civicase__case-card__info">
```
Comment (deb1990) [inline ang/civicase/CaseCard.html]: Better to use `civicase__case-card__activity-info` here

File: scss/components/_checkbox.scss (hunk 7 above)
[the remark this answers is not in the document]
Comment (deb1990) [reply scss/components/_checkbox.scss]: Ok lets keep it here for now.

File: ang/civicase/CaseCard.html (hunk 10 above)
Comment (swastikpareek) [reply ang/civicase/CaseCard.html]: Sorry, I forgot to mention. Panel class is adding extra border-radius of 2px and margin-bottom of 20px which we don't need in the styles. That's why we are not using panel class here.

File: ang/civicase/CaseCard.html [hunk 16]
```diff
@@ -0,0 +1,109 @@
...
+      </div>
+      <div class="civicase__case-card__activity-count-container">
+        <span>Comms: </span>
+        <span class="civicase__case-card__activity-count civicase__case-card__activity-count--zero">0</span>
+      </div>
+      <div class="civicase__case-card__activity-count-container">
+        <span>Other: </span>
+        <span class="civicase__case-card__activity-count">1</span>
+        <span class="civicase__pipe"> | </span>
+        <span class="civicase__case-card__other-count civicase__overdue-activity-icon civicase__overdue-activity-icon--count">3</span>
```
[the remark this answers is not in the document]
Comment (swastikpareek) [reply ang/civicase/CaseCard.html]: No we don't. Removed.

File: ang/civicase/CaseCard.html (hunk 10 above)
Comment (deb1990) [reply ang/civicase/CaseCard.html]: ok, but we need to follow the styleguide of `panel-strip-primary`. And that includes the panel class. Once that is done, we can override any styles which we dont need.

**Pull request merged**